From 2f84ec6918cc6ad73b0c51cedd623ecdf06f085a Mon Sep 17 00:00:00 2001
From: Colin Patrick Mccabe <cmccabe@cloudera.com>
Date: Tue, 23 Feb 2016 11:11:22 -0800
Subject: [PATCH 1261/2848] HADOOP-12829. StatisticsDataReferenceCleaner
 swallows interrupt exceptions (Gregory Chanan via
 cmccabe)

(cherry picked from commit d9c409a4286e36387fb39e7d622e850c13315465)
(cherry picked from commit 7dfcf403525af77763bede12a6c3db91a59252ce)

Change-Id: I4b3b886bfda7d24c15e94bbb9e7cef1567c552b0
---
 .../main/java/org/apache/hadoop/fs/FileSystem.java |    9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
index 70a29af..0995b0a 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
@@ -3088,15 +3088,16 @@ public void cleanUp() {
     private static class StatisticsDataReferenceCleaner implements Runnable {
       @Override
       public void run() {
-        while (true) {
+        while (!Thread.interrupted()) {
           try {
             StatisticsDataReference ref =
                 (StatisticsDataReference)STATS_DATA_REF_QUEUE.remove();
             ref.cleanUp();
+          } catch (InterruptedException ie) {
+            LOG.warn("Cleaner thread interrupted, will stop", ie);
+            Thread.currentThread().interrupt();
           } catch (Throwable th) {
-            // the cleaner thread should continue to run even if there are
-            // exceptions, including InterruptedException
-            LOG.warn("exception in the cleaner thread but it will continue to "
+            LOG.warn("Exception in the cleaner thread but it will continue to "
                 + "run", th);
           }
         }
-- 
1.7.9.5

